CMake: Fix ARMClang compiler issues

urutva opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Update arm compiler common flags which was causing build failures.
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

urutva committed:
cmake: Stringify develop.cmake

Cleanup develop.cmake to remove ARMC5 and uArm compiler flags and
stringify cmake variables.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
cmake: Using Arm compiler from Mbed Studio

Added a command line option `-DMBEDIDE=ON` which allows using Arm
compiler shipped with Mbed Studio on local machines.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater commented on cmake/profiles/develop.cmake:
```diff
@@ -34,69 +34,27 @@ set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} \
-    --c99
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} \
+    -std=gnu11 \
+"
```
Our style seems to prefer the `")` on the same line together, rather than separated.

Patater commented on cmake/app.cmake:
```diff
@@ -2,10 +2,10 @@
 include(${MBED_ROOT}/cmake/core.cmake)
 include(${MBED_ROOT}/cmake/profile.cmake)
+include(${MBED_ROOT}/cmake/env.cmake)
+include(${MBED_ROOT}/cmake/util.cmake)
```
The commit message "cmake: Fix arm compiler build failures" doesn't mention why the re-ordering needs to take place. It only mentioned updating arm compiler common flags. Did we mean to re-order here?

urutva replied:
@Patater Created to new commit explaining the reason for reorder.

Patater commented on cmake/cores/Cortex-M4F.cmake:
```diff
@@ -38,10 +40,10 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
 ")
 set(CMAKE_ASM_FLAGS "${CMAKE_ASM_FLAGS} \
-    ${ARM_FLAGS} \
+    --cpu=Cortex-M4 \
```
Why shouldn't ASM use ARM_FLAGS?

urutva replied:
Updated commit message.

Patater commented on cmake/cores/Cortex-M4F.cmake:
```diff
@@ -38,10 +40,10 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
 ")
 set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} \
-    ${ARM_FLAGS} \
+    --cpu=Cortex-M4 \
```
Why shouldn't the linker use ARM_FLAGS?

urutva replied:
Updated commit message.

Patater commented on cmake/toolchains/ARM.cmake:
```diff
@@ -16,7 +16,6 @@ set(COMMON_FLAGS " \
     -Wno-deprecated-register \
     -fdata-sections \
     -fno-exceptions \
-    -MMD \
```
Why remove MMD? Specify why in commit message.

urutva replied:
Updated commit message.

Patater commented on cmake/toolchains/ARM.cmake:
```diff
@@ -27,10 +26,8 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
-")
-set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} \
-    ${COMMON_FLAGS} \
+    --cpreproc \
```
Why should ASM not use COMMON_FLAGS?

Patater replied:
Should `-c` be in COMMON_FLAGS? Seems to be useful for CFLAGS and CXXFLAGS, but maybe not ASM_FLAGS.

urutva replied:
Updated commit message.

Patater commented on cmake/toolchain.cmake:
```diff
@@ -1,4 +1,8 @@
 
+option(MBEDIDE "Use Arm compiler from Mbed Studio" OFF)
+if(MBEDIDE)
+    set(MBED_STUDIO_ARM_COMPILER "--ide=mbed")
```
What happens if this is passed to a version of Arm compiler that isn't from Mbed Studio?

urutva replied:
It is a command line option which is disabled by default. It is expected to be used to compile using Arm compiler shipped with Mbed studio. If this option is used by mistake then Arm compiler will through a warning message indicating unused command line option.

Patater replied:
OK, so we have to avoid passing this to skip the warning option.

Patater commented on cmake/app.cmake:
```diff
@@ -64,9 +64,9 @@ if(MBED_TOOLCHAIN STREQUAL "GCC_ARM")
+        COMMAND ${ELF2BIN} ${MBED_STUDIO_ARM_COMPILER} --bin  -o $<TARGET_FILE:app>.bin $<TARGET_FILE:app>
         COMMAND ${CMAKE_COMMAND} -E echo "-- built: $<TARGET_FILE:app>.bin"
-        COMMAND ${ELF2BIN} --i32combined  -o $<TARGET_FILE:app>.hex $<TARGET_FILE:app>
+        COMMAND ${ELF2BIN} ${MBED_STUDIO_ARM_COMPILER} --i32combined  -o $<TARGET_FILE:app>.hex $<TARGET_FILE:app>
```
Is Mbed Studio going to use the CMake build system? Is that why the commit "cmake: Add support for Mbed Studio Arm compiler" is needed? Please put the reasoning in the commit message.

urutva replied:
Updated commit message.

urutva committed:
cmake: Re-order cmake script inclusions

The script `cmake/env.cmake` enables languages and in the case of Arm
compiler, `CMAKE_SYSTEM_PROCESSOR` is expected to be set before a
language is enabled. `CMAKE_SYSTEM_PROCESSOR` is set in
`cmake/core.cmake`. Therefore, `cmake/core.cmake` is included before
`cmake/env.cmake`.

Similarly, `cmake/profile.cmake` sets compiler options, hence it is
included before enabling languages in `cmake/env.cmake`.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
cmake: Fix arm compiler build failures

Update arm compiler flags which were causing build failures.

Remove `-MMD` flag as it is not recognised by the Arm compiler.

There are no common flags between compiler, assembler and linker. The
compiler expects `-mcpu` whereas assembler and linker expects `--cpu`.

The `COMMON_FLAGS` in `cmake/toolchains/ARM.cmake` are not recognised by
assembler and linker.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater reviewed (approved):
LGTM

0xc0170 commented on cmake/cores/Cortex-M4F.cmake:
```diff
@@ -38,10 +40,10 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} \
 ")
 set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} \
-    ${ARM_FLAGS} \
+    --cpu=Cortex-M4 \
```
are these changes relevant only to Cortex-M4F (the rest of cores are good?) ? they include `ARM_FLAGS` as I checked some of other cores

urutva replied:
@0xc0170 The `-mfpu` is relevant for cores with `F` appended to its name. But assembler and linker flags needs to be updated for all other cores.
Can we merge this PR as it works with K64F which is the chosen platform?

I'll create another PR with changes needed in other core cmake script.

0xc0170 replied:
yes, i was just checking if we still got a work to do
